go/protocols/materialize/op_future.go

```diff
@@ -0,0 +1,79 @@
+package materialize
+
+import (
+	"fmt"
+
+	pf "github.com/estuary/flow/go/protocols/flow"
+	"go.gazette.dev/core/broker/client"
+)
+
+// OpFuture represents an operation which is executing in the background. The
+// operation has completed when Done selects. Err may be invoked to determine
+// whether the operation succeeded or failed.
+// This is inspired by gazette's `client` package.
+type OpFuture interface {
+	// Done selects when the operation's background execution has finished. The channel
+	// also includes an optional ConnectorState to be emitted as part of the
+	// Acknowledge response
+	Done() <-chan *pf.ConnectorState
```

Comment:
@williamhbaker would that not mean that we cannot calculate the state to be emitted as part of ack in the async operation itself? With the current setup, we return the ack-response state in OpFuture itself and return it. Am I missing something?

Reply:
It would mean that indeed. From what I'm seeing with the current use case that wouldn't be a problem though, since the state for the ack is a checkpoint with all of the currently active bindings being essentially zero'd out, and that can be computed elsewhere. For example, this whole block does not depend on the result of applying the checkpoint:

```go
var checkpointClear = checkpoint{Queries: make(map[string][]string), ToDelete: make(map[string][]string)}
for _, b := range d.bindings {
	checkpointClear.Queries[b.target.Identifier] = []string{}
	checkpointClear.ToDelete[b.target.Identifier] = []string{}
}
checkpointJSON, err = json.Marshal(checkpointClear)
if err != nil {
	return nil, fmt.Errorf("creating checkpoint clearing json: %w", err)
}
```

This pattern of the ack state being something that is pretty easy to pre-compute seems reasonable to me to keep things simpler (IMO). But LMK if there's a requirement to be able to use the result of the "normal" application of the checkpoint to compute the ack checkpoint.

Reply:
@williamhbaker  I think this is quite unintuitive and might give the wrong idea about how the process works. That even though Ack response is emitted after the async operation (and only if the async operation has succeeded), we would in the library want the developer to emit it before the Ack response is even formed. I think the current factoring is a more direct representation of what actually happens.

In the current use case we don't _need_ to emit the state as part of the async operation, but I'm not sure if we might need it when switching to Snowpipe or for other connectors that want to use this behaviour